Added support for the JSON format used by PIXI.js and TexturePacker

limikael opened the pull request:
I added support for a JSON format commonly used for spritesheets in games. I don't know where it comes from originally, but it is at least used by [TexturePacker](https://www.codeandweb.com/texturepacker), [Flash](https://helpx.adobe.com/flash/using/create-sprite-sheet.html) and [PIXI.js](http://www.pixijs.com/).

I have tested a generated spritesheet and it worked to load into PIXI.js. I haven't created a test for unit testing, and in fact it seems like I did something to break it because I use `require('path')`. I think that's why.

So maybe not ready to be merged as it is, but a start!

Any clue why the `require('path')` breaks the tests?

twolfson commented on lib/spritesheet-templates.js:
```diff
@@ -315,6 +315,8 @@ templater.addTemplate('json', require(templatesDir + '/json.template.js'));
 templater.addTemplate('jsonArray', require(templatesDir + '/json_array.template.js'));
 templater.addTemplate('css', require(templatesDir + '/css.template.js'));
 
+templater.addTemplate('jsonTexture', require(templatesDir + '/json_texture.template.js'));
```
`jsonArray` is camelCased due to legacy reasons. We typically go with `snake_case` for template names now (e.g. `scss_maps`). Can we update this name to `json_texture`?

twolfson commented on lib/templates/json_texture.template.js:
```diff
@@ -0,0 +1,43 @@
+path = require("path");
```
You are missing a `var path = require('path');` here. Additionally, we use single quotes for this repo (as enforced by our linter.

```js
var path = require('path');
```

twolfson commented on lib/templates/json_texture.template.js:
```diff
@@ -0,0 +1,43 @@
+  var spriteObj = {};
+
+  // Create frame data for each sprite.
+  spriteObj.frames={};
```
Can we please fix up the spacing here to be consistent with the rest of the repo?

```js
spriteObj.frames = {};
```

twolfson commented on lib/templates/json_texture.template.js:
```diff
@@ -0,0 +1,43 @@
+  // Create frame data for each sprite.
+  spriteObj.frames={};
+
+  data.sprites.forEach(function (sprite) {
```
Can we please name this inline function for better stack traces?

```js
data.sprites.forEach(function saveSprite (sprite) {
```

twolfson commented on lib/templates/json_texture.template.js:
```diff
@@ -0,0 +1,43 @@
+        "x": sprite.x,
+        "y": sprite.y,
+        "w": sprite.width,
+        "h": sprite.height
```
The linter should enforce that we don't need to quote keys unless they aren't valid identifiers otherwise (e.g. have a `-` or are reserved).

```js
        x: sprite.x,
        y: sprite.y,
        w: sprite.width,
        h: sprite.height
```

twolfson commented on lib/templates/json_texture.template.js:
```diff
@@ -0,0 +1,43 @@
+      }
+    };
+
+    spriteObj.frames[path.basename(sprite.source_image)]=entry;
```
Can we please use `sprite.name` instead of `path.basename(sprite.source_image)`? This allows for adjustment via `grunt-spritesmith's cssVarMap`. Additionally, please make the spacing consistent with the rest of this repo:

```js
spriteObj.frames[sprite.name] = entry;
```

twolfson commented on lib/templates/json_texture.template.js:
```diff
@@ -0,0 +1,43 @@
+  spriteObj.frames={};
+
+  data.sprites.forEach(function (sprite) {
+    var entry={
```
Please make the spacing consistent with the remainder of the repo:

```js
var entry = {
```

twolfson commented on lib/templates/json_texture.template.js:
```diff
@@ -0,0 +1,43 @@
+  spriteObj.meta = {
+    "app": "spritesheet-templates",
+    "version": "9.5.0",
+    "image": path.basename(data.spritesheet.image),
```
Is there ever a time when we would have a full path for the `image`?

twolfson commented on lib/templates/json_texture.template.js:
```diff
@@ -0,0 +1,43 @@
+    "app": "spritesheet-templates",
+    "version": "9.5.0",
+    "image": path.basename(data.spritesheet.image),
+    "format": "RGBA8888",
```
Is there ever a time when we wouldn't have an alpha channel nor 32-bit sprite? (e.g. `png-8`, `jpeg`)

twolfson commented on lib/templates/json_texture.template.js:
```diff
@@ -0,0 +1,43 @@
+  // Create the meta data.
+  spriteObj.meta = {
+    "app": "spritesheet-templates",
+    "version": "9.5.0",
```
Please pull in version from our package.json:

```
// At the top of this file
var pkg = require('../../package.json');

// Inline here:
version: pkg.version,
```

twolfson replied:
Please use single quote for object values. It is enforced by the linter and gives this repo consistency.

```js
  app: 'spritesheet-templates',
  version: '9.5.0',
```

twolfson commented on lib/templates/json_texture.template.js:
```diff
@@ -0,0 +1,43 @@
+    "scale": 1,
+    "size": {
+      "w": data.spritesheet.width,
+      "h": data.spritesheet.height
```
You will need to remove these quotes to get the linter passing. It enforces no quotes if the object key is a valid identifier without them.

```js
spriteObj.meta = {
  app: 'spritesheet-templates',
  version: '9.5.0',
  image: path.basename(data.spritesheet.image),
  format: 'RGBA8888',
  scale: 1,
  size: {
    w: data.spritesheet.width,
    h: data.spritesheet.height
  }
};
```

twolfson commented:
Thanks for PR! I am all for supporting this format. Outside of the inline comments, I have a few more requests/questions:

- Is there ever a time when we will need retina support? If so, what does that look like for TexturePacker and friends?
- Can we add tests to the PR? I have a pretty good idea of what the format looks like but an `expected_files` usually clarifies it directly.

twolfson commented:
One more question: From the TexturePacker documentation, I see `.tpsheet` as the typical extension for TexturePacker's JSON data. Is there a common file extension used by TexturePacker, Flash, and pixi.js?

https://www.codeandweb.com/texturepacker/documentation#first-sheet

Also, notes for self. Here are some snippets from where pixi.js interacts with our TexturePacker format:

http://www.goodboydigital.com/pixijs/docs/files/src_pixi_loaders_SpriteSheetLoader.js.html#l5

http://www.goodboydigital.com/pixijs/docs/files/src_pixi_loaders_JsonLoader.js.html#l5

limikael commented:
Thanks for the comments! Will work to tick them off one by one, it will not happen today but will happen in the near future...

Some questions and comments though...

What does "retina support" actually mean in this context? Does it mean that the images will be scaled to the double size? And how does one test this?

Where do you see the documentation about the .tpsheet extension? I checked:

https://www.codeandweb.com/texturepacker/documentation#first-sheet

And it doesn't mention .tpsheet anywhere on that page. It mantions .tps, but that's the extension of the project files for TexturePacker, not the generated sprite sheets. Also googled .tpsheet but didn't find that much information. But if this is the extension that is used then we should use it too I guess.

twolfson commented:
No worries about timeline. When you are done with all the comments, please post another comment here. Unfortunately, GitHub doesn't notify me when there are new commits =(

# Retina support
Retina support means handling a normal and retina spritesheet from the same sprite definition. To make sure we are on the same page, this might be some redundant information. With newer laptops with high pixel density screens, some images can appear granular/pixelated. This is due to normal images that are drawn at a normal ppi (e.g. 100 pixels drawn over 1 inch) are being scaled across longer distances (e.g. 200 pixels over 1 inch).

In order to compensate for this, we are generating a secondary spritesheet which is twice as large yet when used is scaled down the normal size (e.g. it would be 200 pixels over 2 inches, but we are scaling so it's 200 pixels over 1 inch). This is done via attributes like `background-size`.

https://github.com/twolfson/spritesheet-templates/blob/9.5.0/test/expected_files/css_retina.css

```
END OF RETINA EXPLANATION
```

Since pixi.js is HTML based I am curious if it has/needs retina spritesheet support. There seems to be a resolved issue on this and I think the `scale` attribute might be the one to solve it with respect to our template. However, I believe you have more context to pixi.js' usage which is why I would like your insight:

https://github.com/GoodBoyDigital/pixi.js/issues/621

To test out that a retina spritesheet is working, in the CSS context we can remove the `media` query clause to verify that scaling displacement work fine. Retina spritesheets are compatible with all platforms but considered a waste of bandwidth since they add nothing for non-high pixel dense displays.

```css
.icon-sprite-dash-case {
  background-image: url(nested/dir/spritesheet.png);
  background-position: 0px 0px;
  width: 10px;
  height: 20px;
}

/* @media (-webkit-min-device-pixel-ratio: 2),
       (min-resolution: 192dpi) { */
  .icon-sprite-dash-case {
    background-image: url(nested/dir/spritesheet-2x.png);
    background-size: 80px 100px;
  }
/* } */
```

# .tpsheet extension
I saw the `.tpsheet` extension in the screenshot.

https://cdn.codeandweb.com/img/tp-documentation/spritesheet-mac.png

![spritesheet-mac](https://cloud.githubusercontent.com/assets/902488/7795063/cf720f8e-0296-11e5-8b1b-b3335c7bf201.png)

I think the name for `spritesheet-templates` is fine. I was asking about it for auto detection in `grunt-spritesmith` and `gulp.spritesmith`.

https://github.com/Ensighten/grunt-spritesmith/blob/4.6.1/src/grunt-spritesmith.js#L39-L46

limikael commented:
Ok now it's cleaned up! To summarize what I did:

* The name is now `json_texture` instead.
* Use var declaration for the require statement, so the tests can run.
* Fixed things such as double vs. single quotes and removed quotes for object keys, so the linter is now happy.
* Named the inline function.
* Use `sprite.name` for naming the keys. However, see comment 3 below!
* Don't use basename() when referring to the generated image. I realized that the path resolution for the image file is really clever and functional.
* Removed the format specification that said "RGBA8888". It doesn't seem to bother PIXI.js, and it is better to give no information at all than possibly wrong information.
* Load proper package version rather than hard code it.
* I created a unit test for it.

Some comments:

1. Regarding file extension, I don't think there is anything more standard than `.json` unfortunately. The screenshot from TexturePacker is for another format, if you look just above the file name input field you see that it says "Unity - Texture 2D Sprite".
2. I don't think it makes sense to provide a retina option. PIXI is used together with a canvas that usually covers the whole screen. The scaling of the images is usually done using PIXI, so what units the browser use doesn't really matter. I could be wrong though, but I think we could leave it like it is and let the future show if there is any need for it.
3. The default behaviour for TexturePacker is to use the full filename for the keys, which is why I had the code `path.basename(sprite.source_image)` there. I replaced it with just `sprite.name` to be more standard compliant. This means that generated spritesheets will not work by default as expected for someone who comes from TexturePacker. It should be documented that they need to use this function in their Gruntfile.js:

```
cssVarMap: function(sprite) {
    sprite.name = path.basename(sprite.source_image);
}
```

As a suggestion, could we make the cssVarMap accept either a function (as it does now), or a reference to a set of built in functions? Maybe also an array with such names? So instead of the code above, users could have:

    cssVarMap: "imageBasename"

or:

    cssVarMap: ["imageBasename", function(sprite) {
        // do something else user defined with the definition 
    })]

But this would go in the grunt-spritesmith project of course...

Let me know what you think!

twolfson commented on test/json_texture_test.js:
```diff
@@ -0,0 +1,28 @@
+var fs = require('fs');
+var pkg = require('../package.json');
+
+describe('A hash map of texture info, similar to TexturePacker', function () {
```
In JS, we refer to a hash map as an object. In this case though, `configUtils.multipleSprites` is an array =/

limikael replied:
Ok, I'll call it 'An object map of texture info, similar to TexturePacker'? What I'm referring to is what it is outputting, I don't really understand what `configUtils.multipleSprites` has to do with it?

twolfson replied:
From the perspective of BDD, we are referring to the initial context which is information about multiple sprites. We don't start talking about the generated result until we reach "process by `spritesheet-templates`".

limikael replied:
gotcha

twolfson commented on test/json_texture_test.js:
```diff
@@ -0,0 +1,28 @@
+var configUtils = require('./utils/config');
+var testUtils = require('./utils/test');
+var fs = require('fs');
+var pkg = require('../package.json');
```
While it might not seem obvious, we have a sort order for our dependencies:

1. Core library (e.g. `assert`, `fs`)
2. Third party libraries (e.g. `request`, `grunt`)
3. First party libraries (e.g. `../package.json`, `./utils/config`)

```js
var assert = require('assert');
var fs = require('fs');
var configUtils = require('./utils/config');
var testUtils = require('./utils/test');
var pkg = require('../package.json');
```

limikael replied:
Ok fixing that...

twolfson commented on test/json_texture_test.js:
```diff
@@ -0,0 +1,28 @@
+        JSON.parse(result);
+      });
+    });
+  }
```
We can prob relocate the shared `assertValidJson` to `testUtils`

limikael replied:
don't really understand what exactly you want me to do. I'll check it in the way it is, if there are things you want to move around maybe you can do that after the merge?

twolfson replied:
This `assertValidJson` also exists in the JSON test file. I was suggesting to relocate the assertions into `utils/test.js` for reused cose. This is fine though, I will move things around post merge.

twolfson commented on test/json_texture_test.js:
```diff
@@ -0,0 +1,28 @@
+
+    var expected = fs.readFileSync(__dirname + '/expected_files/json_texture.json', 'utf8');
+    expected = expected.replace('__VERSION__', pkg.version);
+    testUtils.assertOutputMatchesString(expected);
```
I find it weird that I never moved to `deepEqual` comparison for JSON files. However, this current setup is always reading the expected file outside of `mocha's` context (e.g. not in a `before`/`after`/`it`). We should consider removing the test utility and placing it inline for this test:

```js
testUtils.runTemplater({format: 'json_texture'});

it('matches as expected', function () {
  var expected = fs.readFileSync(__dirname + '/expected_files/json_texture.json', 'utf8');
  expected = expected.replace('__VERSION__', pkg.version);
  assert.strictEqual(this,result, expected);
});
```

limikael replied:
ok putting it inline

twolfson commented:
Whoo, great turn around time and the changes look good. There are a few minor one-offs that I left inline comments on.

With respect to your comments:

- Since it sounds like `.tpsheet` is not a canonical extension, we will avoid the auto-detection and require `cssFormat: 'json_texture'` for now.
- I appreciate your insight on retina support. Keeping it as is works for me as well =)
- As for naming of the `frames`, it's usually best to keep with convention. I think going back to the original `path.basename` will work for now.
    - If we encounter edge cases, then here are some early options:
        - Add a formatter option which indicates we want to use `sprite.name` for the frame key instead
        - Add new template variable like `sprite.basename` which would be generated by `spritesmith` and used by our template
        - Ask users to modify `source_image` over `name` via `cssVarMap`

limikael commented:
Ok made some fixes as suggested in the comments. The name resolution works like this:

1. It checks if there is an option called frame_name. If so it uses that. This keeps the possibility to change the name of the sprite.
2. It checks if there is a source_image option. If so, it uses path.basename(sprite.source_image)
3. If not, it will use sprite.name

I added frame_name because asking users to modify sprite.source_image wouldn't really work always, if you would want to add something like an extra directory, because that would be stripped away by path.basename.

twolfson commented:
This looks great. I really like the `frame_name` decision :+1:

I will land this by the end of the weekend. Thanks for this awesome PR and your quick replies =)

limikael commented:
Cool! Wrote some documentation. I referenced grunt-spritesmith from it, as you will see. Don't know if you want to do that or if you want to see them as being on different levels in some dependency hierarchy. So feel free to edit that!

twolfson commented:
Ah, I was planning on writing docs while landing. Oh well, this works too. It looks good but I will make a few edits during landing (e.g. making it more concise).

twolfson commented:
This has been landed/released in `9.6.0`. The corresponding `grunt-spritesmith`/`gulp.spritesmith` releases should go out in a few minutes. Once again, thanks for the PR! =)
